### ISSUE TITLE
Out of sync and crashes when performing maintenance

### ISSUE THREAD

**Issue by andrenarchy** [stellar/stellar-core#1725]
# Issue type
bug

# Version found
9.2.0

# Your environment and setup
Running on Docker in a Kubernetes cluster.
```
Linux stellar-mainnet-3-stellar-core-54f64c996-xvsgg 4.14.22+ #1 SMP Wed May 30 13:49:27 PDT 2018 x86_64 GNU/Linux
```

# Issue description
We observe that our validator nodes get out of sync frequently and that the process even terminates sometimes. Looking at the logs it seems like the last message before this is always a `performing maintenance` message, max 2 minutes before the node goes out of sync (`Out of sync context` message). Then it usually only takes 1 second after the out of sync message and the process terminates (`Application destructing` message).

## Supporting files
Excerpt from the logs with maintenance, out-of-sync, and destructing messages:
```
2018-07-04T01:53:46.324 GAK6Z [default INFO] Performing maintenance
2018-07-04T03:14:38.593 GBJQU [default INFO] Performing maintenance
2018-07-04T03:17:23.031 GC5SX [default INFO] Performing maintenance
2018-07-04T04:17:11.509 GCLWF [default INFO] Performing maintenance
2018-07-04T05:54:43.693 GAK6Z [default INFO] Performing maintenance
2018-07-04T05:55:22.592 GAK6Z [Herder INFO] Out of sync context: {
2018-07-04T05:55:23.767 GAK6Z [default INFO] Application destructing
2018-07-04T07:15:35.920 GBJQU [default INFO] Performing maintenance
2018-07-04T07:17:11.172 GBJQU [Herder INFO] Out of sync context: {
2018-07-04T07:18:23.051 GC5SX [default INFO] Performing maintenance
2018-07-04T07:19:19.032 GC5SX [Herder INFO] Out of sync context: {
2018-07-04T07:19:20.082 GC5SX [default INFO] Application destructing
2018-07-04T08:17:11.829 GCLWF [default INFO] Performing maintenance
2018-07-04T09:55:24.209 GAK6Z [default INFO] Performing maintenance
2018-07-04T11:16:36.157 GBJQU [default INFO] Performing maintenance
2018-07-04T11:19:20.615 GC5SX [default INFO] Performing maintenance
2018-07-04T12:17:12.110 GCLWF [default INFO] Performing maintenance
2018-07-04T13:56:23.849 GAK6Z [default INFO] Performing maintenance
2018-07-04T15:17:36.178 GBJQU [default INFO] Performing maintenance
2018-07-04T15:20:12.956 GC5SX [default INFO] Performing maintenance
2018-07-04T16:17:12.533 GCLWF [default INFO] Performing maintenance
2018-07-04T17:57:23.835 GAK6Z [default INFO] Performing maintenance
2018-07-04T19:18:36.251 GBJQU [default INFO] Performing maintenance
2018-07-04T19:19:32.780 GBJQU [Herder INFO] Out of sync context: {
2018-07-04T19:19:33.818 GBJQU [default INFO] Application destructing
2018-07-04T19:21:13.045 GC5SX [default INFO] Performing maintenance
2018-07-04T20:17:12.799 GCLWF [default INFO] Performing maintenance
2018-07-04T21:58:23.698 GAK6Z [default INFO] Performing maintenance
2018-07-04T23:19:34.089 GBJQU [default INFO] Performing maintenance
2018-07-04T23:22:12.980 GC5SX [default INFO] Performing maintenance
2018-07-04T23:23:10.234 GC5SX [Herder INFO] Out of sync context: {
2018-07-04T23:23:11.280 GC5SX [default INFO] Application destructing
2018-07-05T00:17:13.083 GCLWF [default INFO] Performing maintenance
2018-07-05T01:59:23.887 GAK6Z [default INFO] Performing maintenance
2018-07-05T03:20:26.172 GBJQU [default INFO] Performing maintenance
2018-07-05T03:23:11.679 GC5SX [default INFO] Performing maintenance
2018-07-05T04:17:13.345 GCLWF [default INFO] Performing maintenance
2018-07-05T06:00:23.787 GAK6Z [default INFO] Performing maintenance
2018-07-05T06:01:03.853 GAK6Z [Herder INFO] Out of sync context: {
2018-07-05T06:01:04.920 GAK6Z [default INFO] Application destructing
2018-07-05T07:21:26.294 GBJQU [default INFO] Performing maintenance
2018-07-05T07:24:13.134 GC5SX [default INFO] Performing maintenance
```

**Comment by MonsieurNicolas** [comment]
This may be caused by having a large backlog for the maintenance queue and the fact that postgresql doesn't deal very well with sparse tables.

One way to get around this is to perform the following:
1. perform the maintenance on your validator manually, running the command `maintenance?queue=true&count=50000000` (the 50000000 number is just a very high number that exceeds the number of ledgers in the database). This may take up to an hour depending on your database. This will delete old entries from the database.
2. perform a `VACUUM FULL` command the stellar-core postgres database. This will also take a while to run and will cause the database to shrink on disk a lot (we've seen our nodes go from 90GB down to 1GB).

Let me know if this works for you @andrenarchy

**Comment by andrenarchy** [comment]
@MonsieurNicolas Thanks, I could try those but in my opinion it'd be great if this could be fixed in a way that doesn't require manual interaction. Is there anything I can do to verify that it's the maintenance queue and could it be implemented in a way that doesn't interfere with the validator's sync state?

**Comment by andrenarchy** [comment]
So I just got hold of some DEBUG log files when it went out of sync: [stellar3.log](https://github.com/stellar/stellar-core/files/2170928/stellar3.new.log). It did maintenance before again but I don't get the logs from before the beginning of this log file out easily (it's an issue we're figuring out with google cloud atm). However, this log file contains the `Out of sync` message, maybe this helps?

**Comment by MonsieurNicolas** [comment]
afaik it's a limitation of the way postgresql performs vacuum: the background process doesn't actually do much and the unused space is not reclaimed. I am not a postgresql expert, there might be a better way. 
Relevant section of the documentation is https://www.postgresql.org/docs/10/static/sql-vacuum.html : 
> Plain VACUUM (without FULL) simply reclaims space and makes it available for re-use. This form of the command can operate in parallel with normal reading and writing of the table, as an exclusive lock is not obtained. However, extra space is not returned to the operating system (in most cases); it's just kept available for re-use within the same table. VACUUM FULL rewrites the entire contents of the table into a new disk file with no extra space, allowing unused space to be returned to the operating system. This form is much slower and requires an exclusive lock on each table while it is being processed.

After performing this maintenance, the performance should be good as maintenance will keep the database from growing (ie, there is no difference between the amount of data produced and the amount of data being deleted by maintenance).

Basically, what I've seen on some of our nodes is that before `VACUUM FULL`, maintenance would lock the database for long period of time (sometimes in the order of 15 minutes), which causes the node to fall out of sync; after `VACUUM FULL`, the same type of maintenance takes less than a second.

Note that this is a "one time" thing: the reason we need this aggressive vacuuming is because after performing a large catchup, the "txhistory" table grows to contain a large part of history. When catchup is complete, data is published to history archive and horizon ingested the data then maintenance can start doing its work. When that happens rows in txhistory are deleted and at some point the postgresql engine cannot keep its indexes in sync (as far as I can tell), and turns indexed queries into linear scans.

Only workaround I found so far is to perform this full maintenance after all data has been ingested/published as outlined earlier.

The only improvements I can see are:
* parallelize "publish" and catchup (not implemented yet)
  * by publishing to history archive (assuming horizon can also ingest data as it gets produced), maintenance can keep the number of active rows to roughly a constant number (behave the same way that in normal steady state).
  * I think that for this to work well, we'll have to halt catchup if publishing and maintenance can't keep up. Basically, we'll make the unit of work the sequence "wait for cursors to be within X ledgers", "replay snapshot",  "publish snapshot", "maintenance". 
* disable maintenance can be a workaround, but that means that the core database will keep growing (and core's database is not meant to be used for historical data)

**Comment by andrenarchy** [comment]
So even with manually sending the `maintenance` command (as described above), we still have the problem that the core process exits/crashes a couple of seconds/minutes after sending the command. This definitely looks like a bug to me.

@MonsieurNicolas any other idea about how we can get the maintenance done?

**Comment by andrenarchy** [comment]
This is 100% reproducible now. I just made Core run in DEBUG log mode but after sending the maintenance command the only thing I see in the logs is this before the process exits
```
2018-07-09T12:41:38.318 GAK6Z [default INFO] Performing maintenance
2018-07-09T12:41:38.320 GAK6Z [History INFO] Trimming history <= ledger 18831663 (rmin=18831663, qmin=18833462, lmin=18833398)
```
The postgres instance logs this:
```
LOG:  unexpected EOF on client connection with an open transaction
```

**Comment by andrenarchy** [comment]
One more log file from a crash right now: 
[stellar3.2018-07-10T07:46:00.log](https://github.com/stellar/stellar-core/files/2179338/stellar3.2018-07-10T07.46.00.log). 

The last bit is interesting:
```
2018-07-10T07:44:09.947 GAK6Z [Overlay INFO] New connected peer (empty)
2018-07-10T07:44:09.948 GAK6Z [default INFO] got signal 15, shutting down
2018-07-10T07:44:09.974 GAK6Z [Overlay ERROR] readHeaderHandler error: End of file :(empty) [TCPPeer.cpp:397]
2018-07-10T07:44:10.974 GAK6Z [default INFO] Application destructing
2018-07-10T07:44:10.977 GAK6Z [default INFO] Application destroyed
```
Why does it get SIGTERM (signal 15)?

**Comment by andrenarchy** [comment]
Alright, I know what's happening now:
1. Core maintenance starts
2. Core gets blocked, i.e., all connections to the HTTP port are hanging
3. Our Kubernetes "liveness probe" checks the HTTP port and the request times out. After a couple of unsuccessful requests it terminates the container and spins up a new one.

So our health checks are responsible for the process termination. However, the root cause is that Core gets blocked completely. And this is also the reason for it getting out of sync. 

@MonsieurNicolas what are your thoughts on this?

**Comment by MonsieurNicolas** [comment]
The incremental/automated "maintenance" will not work properly until you perform full maintenance, including running "vacuum full" on the database. Did you try to follow the instructions I gave earlier?

**Comment by andrenarchy** [comment]
Yes what I described in [this comment](https://github.com/stellar/stellar-core/issues/1725#issuecomment-403771543) also happened when I followed your instructions (that's what I meant with sending the command manually). The process gets killed because it's unresponsive. I had to disable the health checks and then I could complete the full maintenance and run `vacuum full`.

Even though this is a one time thing, it's not intuitive at all and the fact that the maintenance is run inside the Core process which then stops responding is making things even more complicated. What can we do to simplify this?

**Comment by MonsieurNicolas** [comment]
Sorry, I missed your response.
As this is an administrative command, you should either not run this command from within kubernetes, or easier might be to disable the probe while you perform the command

**Comment by andrenarchy** [comment]
Yeah, we disabled the probe temporarily. It's just a bit annoying since you can't just spin up a new Stellar node without manually disabling/re-enabling this.

I'm not sure what you mean with not running the command from within kubernetes – that's just where we run the nodes so there's no alternative.

**Comment by MonsieurNicolas** [comment]
> I'm not sure what you mean with not running the command from within kubernetes – that's just where we run the nodes so there's no alternative.

When I wrote this I assumed you were offloading all state off the box (stateless kubernetes), but while it's fine for the database, it's not for the bucket folder.

I am opening a different issue to discuss how to never have to perform this manual maintenance

**Comment by fuzzyami** [comment]
we've run into this issue on various of our nodes: they simply lose sync because auto-maintenance (ie deleting from the db) takes too long. Running the procedure mentioned above usually helps, but in some cases, maintenance (auto or manual) simply does **nothing** - it returns immediately w/o deleting anything. very odd. perhaps you have an idea why this might happen @MonsieurNicolas ? note this happens on cores with ~500K entries in their txhistory table...

In general, I think the 4-hours default is too high and should be reduced if traffic is significant. you'll be erasing (less) data more often, which works around this problem and has no drawbacks, AFAIK.

another tip regarding VACUUM FULL: it does take a long time and requires a lot of feee disc space to complete successfully, as it basically creates copies of every table (and then erases the original).

**Comment by marta-lokhova** [comment]
Closing: maintenance is really fast in recent cores because of the minimal usage of SQL across the codebase.
